### PR TITLE
Refactor the Gradle build

### PR DESCRIPTION
**Pull request by melix**
This commit refactors the Gradle build to use the new Micronaut
build plugins and move it to the new build template (which makes
use of convention plugins).

The build is obviously more complex than for external Micronaut
modules, in particular because some modules are providing core
functionalities but also depend on external Micronaut modules
for tests (e.g Micronaut Test).

### CODE REVIEW

**Comment by sdelamo** [comment]
@timyates please check this PR. I would like to move this forward for 3.4.1

**Review by alvarosanchez — APPROVED** [review]

**Review by timyates — APPROVED** [review]
💪😎

File: buildSrc/build.gradle
Comment (sdelamo) [inline buildSrc/build.gradle]: should not this be 5.3.0?

File: buildSrc/build.gradle
Comment (sdelamo) [inline buildSrc/build.gradle:10]: should we have a `buildSrc/gradle.properties` to have these versions numbers defined there?

File: core/build.gradle
Comment (sdelamo) [inline core/build.gradle:10]: are we removing slf4j from core? would not this be a breaking change?

File: http/build.gradle
Comment (sdelamo) [inline http/build.gradle]: Please, indent this.

File: inject-kotlin-test/build.gradle
Comment (sdelamo) [inline inject-kotlin-test/build.gradle]: please, indent this.

File: session/build.gradle
Comment (sdelamo) [inline session/build.gradle:2]: why is session a core library? I assume you added because `shadowJarEnabled=true` but does it need to be a core library?

File: buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java
Comment (sdelamo) [inline buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: In this PR, you define five modules as core libraries `inject`, `aop`, `core`, `runtime`, `session`.

In addtion to `bom`, this excludes only `runtime` and `inject` but not `core`, `aop` or `session`. Is this correct?

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: buildSrc/build.gradle
Comment (melix) [reply buildSrc/build.gradle]: It wasn't released when I created the PR, will update :)

File: buildSrc/build.gradle
Comment (melix) [reply buildSrc/build.gradle:10]: I don't think it's worth it for now. Actually we should get rid of `buildSrc` in favor of `build-logic` soon. In that process we could reuse the version catalog too.

File: core/build.gradle
Comment (melix) [reply core/build.gradle:10]: That library is added automatically by applying the `base convention plugin (transitively)

File: session/build.gradle
Comment (melix) [reply session/build.gradle:2]: That is simply remodeling what existed before, `session` was treated differently like the other `core-library`. I am not familiar enough with that module to say if it should be done differently.

File: buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java
Comment (melix) [reply buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: That's correct. Again I'm not sure if this is intentional or not: this was what was done before. Note that I have checked that the produced jars, pom and module files are identical _byte-to-byte_ using this PR.

File: inject-kotlin-test/build.gradle
Comment (melix) [reply inject-kotlin-test/build.gradle]: woot, replace in files :)

**Comment by melix** [comment]
The build won't pass: after upgrading to build tools 5.3.0 it now detects binary breaking changes:

image

image

image

File: session/build.gradle
Comment (sdelamo) [reply session/build.gradle:2]: It was added here: https://github.com/micronaut-projects/micronaut-core/pull/1486/files

File: buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java
Comment (sdelamo) [reply buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: > The build won't pass: after upgrading to build tools 5.3.0 it now detects binary breaking changes:

Why would it fail? 

Should the binary check not check if this PR introduces any breaking changes from 3.4.0 to this PR?

Comment (melix) [reply buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: It's not failing because of changes in this PR. It's failing because it upgrades to build plugins 5.3.0, which introduce binary compatibility checks, which fail because they detect binary incompatible changes between 3.4.1-SNAPSHOT and 3.4.0 (see screenshots in comments).

Comment (sdelamo) [reply buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: yes, but those changes have not been introduced between 3.4.0 and 3.4.1-SNAPSHOT.  It seems the binary check is comparing 3.3.4 and 3.4.0 not 3.4.0 and 3.4.1-SNAPSHOT

Comment (melix) [reply buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: No I think the problem was that it wasn't up-to-date with `3.4.x`. The binary checks pass now. However now it's BOM checking which doesn't :D

**Comment by melix** [comment]
For clarity here are the BOM validation problems which are now detected: https://ge.micronaut.io/s/zphruxszgq6d4/console-log#L560-L579

```
BOM io.micronaut.serde:micronaut-serde-bom:1.0.0 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.serde:micronaut-serde-bom:1.0.0) declares a non-resolvable dependency: io.micronaut.serde:micronaut-serde-tck:1.0.0
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-oraclecloud-sdk-processor:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-oraclecloud-atp-hikari-test:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-oraclecloud-atp-ucp-test:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-http-function-java:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-function-groovy:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-http-function-kotlin:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-groovy:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-function-java:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-java:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-http-function-groovy:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-function-kotlin:2.1.1
BOM io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.oraclecloud:micronaut-oraclecloud-bom:2.1.1) declares a non-resolvable dependency: io.micronaut.oraclecloud:micronaut-example-kotlin:2.1.1
BOM io.micronaut.data:micronaut-data-bom:3.3.0 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.data:micronaut-data-bom:3.3.0) declares a non-resolvable dependency: io.micronaut.data:micronaut-data-document-tck:3.3.0
BOM io.micronaut.data:micronaut-data-bom:3.3.0 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.data:micronaut-data-bom:3.3.0) declares a non-resolvable dependency: io.micronaut.data:micronaut-data-tck:3.3.0
BOM io.micronaut.servlet:micronaut-servlet-bom:3.2.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.servlet:micronaut-servlet-bom:3.2.1) declares a non-resolvable dependency: io.micronaut.servlet:micronaut-test-kotlin-jetty:3.2.1
BOM org.jdbi:jdbi3-bom:3.20.1 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.sql:micronaut-sql-bom:4.2.2 -> org.jdbi:jdbi3-bom:3.20.1) declares a non-resolvable dependency: org.jdbi:jdbi3-sqlobject:3.20.1
BOM io.r2dbc:r2dbc-bom:Arabba-SR11 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.r2dbc:micronaut-r2dbc-bom:2.1.0 -> io.r2dbc:r2dbc-bom:Arabba-SR11) declares dependency on com.google.cloud:cloud-spanner-r2dbc:0.6.0 which doesn't belong to group io.r2dbc.
BOM io.r2dbc:r2dbc-bom:Arabba-SR11 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.r2dbc:micronaut-r2dbc-bom:2.1.0 -> io.r2dbc:r2dbc-bom:Arabba-SR11) declares dependency on com.oracle.database.r2dbc:oracle-r2dbc:0.1.0 which doesn't belong to group io.r2dbc.
BOM io.r2dbc:r2dbc-bom:Arabba-SR11 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.r2dbc:micronaut-r2dbc-bom:2.1.0 -> io.r2dbc:r2dbc-bom:Arabba-SR11) declares dependency on org.mariadb:r2dbc-mariadb:1.0.2 which doesn't belong to group io.r2dbc.
BOM io.r2dbc:r2dbc-bom:Arabba-SR11 (via io.micronaut:micronaut-bom:3.4.1-SNAPSHOT -> io.micronaut.r2dbc:micronaut-r2dbc-bom:2.1.0 -> io.r2dbc:r2dbc-bom:Arabba-SR11) declares dependency on dev.miku:r2dbc-mysql:0.8.2.RELEASE which doesn't belong to group io.r2dbc.
```

File: buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java
Comment (sdelamo) [reply buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: If I understand correctly.  We have to fix: 
- [ ] micronaut-r2dbc-bom
- [ ] micronaut-servlet-bom 
- [ ] micronaut-oraclecloud-bom
- [ ] micronaut-data 
- [ ]micronaut-serde-bom

It will involve releasing versions of those modules updating the `micronaut-core` version catalogue with those new versions. Then merge 3.4.x branch against this PR again.  

right?

Comment (sdelamo) [reply buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: If Updating those modules to 5.3.0, it should trigger these errors in their builds. right?

**Comment by melix** [comment]
I have added suppressions for the problems above and created issues in the respective problematic projects.

File: buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java
Comment (melix) [reply buildSrc/src/main/groovy/io/micronaut/build/internal/ext/DefaultMicronautCoreExtension.java:40]: There's nothing we can do for r2dbc (apart from silencing the problem) as far as I understood, unless we want to inline the dependencies ourselves.
For the other BOMs I have created issues in the projects and silenced the problems here so that we can proceed _without_ waiting for new releases.
Yes, migrating those projects to 5.3.0 would trigger errors in their own builds.

**Comment by sdelamo** [comment]
> I have added suppressions for the problems above and created issues in the respective problematic projects.

Ok, after I merge it I will create a PR to remove the suppressions in a different PR. So that we don't forget.

**Comment by melix** [comment]
> Ok, after I merge it I will create a PR to remove the suppressions in a different PR

You'll have to keep the r2dbc one.

**Comment by sdelamo** [comment]
> > Ok, after I merge it I will create a PR to remove the suppressions in a different PR
> 
> You'll have to keep the r2dbc one.

Should not we keep the r2dbc suppression in the 2rdbc module instead?

**Comment by melix** [comment]
> Should not we keep the r2dbc suppression in the 2rdbc module instead?

You need both: validation will fail in the r2dbc module, but it will also fail here since we check transitively.

**Review by sdelamo — APPROVED** [review]

**Pull request merged**
